File: 2020/day1/day1.py
#!/usr/bin/python3

def sum2(numbers, target):
	i = 0
	j = len(numbers) - 1
	k = numbers[i] + numbers[j]

	while k != target:
		if k > target:
			j -= 1
		else:
			i += 1
		k = numbers[i] + numbers[j]

	return numbers[i], numbers[j]

with open('input.txt') as f:
	numbers = [int(i) for i in f.read().splitlines()]
numbers.sort()
target = 2020

num1, num2 = sum2(numbers, target)
print(num1 * num2)

for i in numbers:
	num1, num2 = sum2(numbers, target - i)
	if num1 != i and num2 != i:
		print(i * num1 * num2)
		break

File: 2020/day1/day1.go
package main

import (
	"fmt"
	"io/ioutil"
	"sort"
	"strconv"
	"strings"
)

func sum2(numbers []int, target int) (int, int) {
	i := 0
	j := len(numbers) - 1
	k := numbers[i] + numbers[j]

	for k != target {
		if k > target {
			j--
		} else {
			i++
		}
		k = numbers[i] + numbers[j]
	}

	return numbers[i], numbers[j]
}

func sum3(numbers []int, target int) (int, int, int) {
	for _, n := range numbers {
		num1, num2 := sum2(numbers, target-n)
		if num1 != n && num2 != n {
			return n, num1, num2
		}
	}
	return 0, 0, 0
}

func get_data() []int {
	data, _ := ioutil.ReadFile("input.txt")
	data = data[:len(data)-1]
	lines := strings.Split(string(data), "\n")
	numbers := []int{}
	for _, line := range lines {
		num, _ := strconv.Atoi(line)
		numbers = append(numbers, num)
	}
	sort.Ints(numbers)
	return numbers
}

func main() {
	numbers := get_data()
	target := 2020

	num1, num2 := sum2(numbers, target)
	fmt.Println(num1 * num2)

	num1, num2, num3 := sum3(numbers, target)
	fmt.Println(num1 * num2 * num3)
}
